source/common/stats/stats_impl.cc

```diff
@@ -135,35 +135,33 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
   return false;
 }
 
-RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  uint64_t num_bytes_to_allocate = RawStatData::structSize(name.size());
-  RawStatData* data = static_cast<RawStatData*>(::calloc(num_bytes_to_allocate, 1));
-  if (data == nullptr) {
-    throw std::bad_alloc();
-  }
-  data->checkAndInit(name, num_bytes_to_allocate);
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
```

Comment:
I mentioned this earlier, but the comment got buried under some other changes.  I'm relatively confident that adding truncation in HeapStatDataAllocator is a user-facing change.  Previously, in non hot restart mode, we did not truncate stats.  This should be documented in the release notes.  I'm not sure it makes sense to change this behavior if we ultimately do not want non hot restart stats to be truncated as is suggested in #3965. I could be wrong about the behavior changing, but if so, please point out where they are currently being truncated on master.

Reply:
Reading again I think you are right; it's pretty subtle though. In current master, we truncate when making a RawStatData in RawStatData::truncateAndInit(), but asymmetrically, we truncate at the call-site in ThreadLocalStoreImpl::safeMakeStat() only when calling the fallback.  OK let me fix that; it's a little cleaner to just do this right.